File: src/TemplateModels/JavaBlock.cs
using System;

namespace AutoRest.Java
{
    public class JavaBlock : JavaContext
    {
        private readonly JavaFileContents contents;

        public JavaBlock(JavaFileContents contents)
        {
            this.contents = contents;
        }

        public void Indent(Action indentAction)
        {
            contents.Indent(indentAction);
        }

        public void IncreaseIndent()
        {
            contents.IncreaseIndent();
        }

        public void DecreaseIndent()
        {
            contents.DecreaseIndent();
        }

        public void Text(string text)
        {
            contents.Text(text);
        }

        public void Line(string text, params object[] formattedArguments)
        {
            contents.Line(text, formattedArguments);
        }

        public void Line()
        {
            contents.Line();
        }

        public void Block(string text, Action<JavaBlock> bodyAction)
        {
            contents.Block(text, bodyAction);
        }

        public void JavadocComment(string text)
        {
            contents.JavadocComment(text);
        }

        public void JavadocComment(Action<JavaJavadocComment> commentAction)
        {
            contents.JavadocComment(commentAction);
        }

        public void Return(string text)
        {
            contents.Return(text);
        }

        public void Annotation(params string[] annotations)
        {
            contents.Annotation(annotations);
        }

        public void ReturnAnonymousClass(string anonymousClassDeclaration, Action<JavaClass> anonymousClassBlock)
        {
            contents.ReturnAnonymousClass(anonymousClassDeclaration, anonymousClassBlock);
        }

        public JavaIfBlock If(string condition, Action<JavaBlock> ifAction)
        {
            contents.If(condition, ifAction);
            return new JavaIfBlock(contents);
        }

        public void Lambda(string parameterType, string parameterName, Action<JavaLambda> body)
        {
            contents.Lambda(parameterType, parameterName, body);
        }

        public void Lambda(string parameterType, string parameterName, string returnExpression)
        {
            contents.Lambda(parameterType, parameterName, returnExpression);
        }
    }
}

File: javagen/src/main/java/com/azure/autorest/model/javamodel/JavaBlock.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.autorest.model.javamodel;

import java.util.function.Consumer;

import static com.azure.autorest.util.TemplateUtil.GENERATED_JAVADOC_DESC_END_MARKER;
import static com.azure.autorest.util.TemplateUtil.GENERATED_JAVADOC_DESC_START_MARKER;
import static com.azure.autorest.util.TemplateUtil.GENERATED_JAVADOC_TAG_END_MARKER;
import static com.azure.autorest.util.TemplateUtil.GENERATED_JAVADOC_TAG_START_MARKER;

public class JavaBlock implements JavaContext {
    private JavaFileContents contents;

    public JavaBlock(JavaFileContents contents) {
        this.contents = contents;
    }

    public final void indent(Runnable indentAction) {
        contents.indent(indentAction);
    }

    public final void increaseIndent() {
        contents.increaseIndent();
    }

    public final void decreaseIndent() {
        contents.decreaseIndent();
    }

    public final void text(String text) {
        contents.text(text);
    }

    public final void line(String text, Object... formattedArguments) {
        contents.line(text, formattedArguments);
    }

    public final void line() {
        contents.line();
    }

    public final void block(String text, Consumer<JavaBlock> bodyAction) {
        contents.block(text, bodyAction);
    }

    public final void javadocComment(String text) {
        contents.javadocComment(text);
    }

    public final void javadocComment(Consumer<JavaJavadocComment> commentAction) {
        contents.javadocComment(commentAction);
    }

    @Override
    public final void javadocComment(Consumer<JavaJavadocComment> commentDescriptionAction, Consumer<JavaJavadocComment> commentTagsAction, boolean withGeneratedMarker) {
        if (commentDescriptionAction != null) {
            contents.javadocCommentStart();
            if (withGeneratedMarker) {
                contents.line(GENERATED_JAVADOC_DESC_START_MARKER);
            }
            commentDescriptionAction.accept(new JavaJavadocComment(contents));

            if (withGeneratedMarker) {
                contents.line(GENERATED_JAVADOC_DESC_END_MARKER);
            }

            if (commentTagsAction == null) { // if no tags action, then end the javadoc comment
                contents.javadocCommentEnd();
            } else { // otherwise, a new line will be added and the tags action will start
                contents.line();
            }
        }
        if (commentTagsAction != null) {
            if (commentDescriptionAction == null) {// if no description action, then start the javadoc comment
                contents.javadocCommentStart();
            }
            if (withGeneratedMarker) {
                contents.line(GENERATED_JAVADOC_TAG_START_MARKER);
            }
            commentTagsAction.accept(new JavaJavadocComment(contents));

            if (withGeneratedMarker) {
                contents.line(GENERATED_JAVADOC_TAG_END_MARKER);
            }
            contents.javadocCommentEnd();
        }
    }

    public final void methodReturn(String text) {
        contents.methodReturn(text);
    }

    public final void annotation(String... annotations) {
        contents.annotation(annotations);
    }

    public final void returnAnonymousClass(String anonymousClassDeclaration, Consumer<JavaClass> anonymousClassBlock) {
        contents.returnAnonymousClass(anonymousClassDeclaration, anonymousClassBlock);
    }

    public final void anonymousClass(String anonymousClassDeclaration, String instanceName, Consumer<JavaClass> anonymousClassBlock) {
        contents.anonymousClass(anonymousClassDeclaration, instanceName, anonymousClassBlock);
    }

    public final JavaIfBlock ifBlock(String condition, Consumer<JavaBlock> ifAction) {
        contents.ifBlock(condition, ifAction);
        return new JavaIfBlock(contents);
    }

    public final JavaTryBlock tryBlock(Consumer<JavaBlock> ifAction) {
        contents.tryBlock(ifAction);
        return new JavaTryBlock(contents);
    }

    public final JavaTryBlock tryBlock(String resource, Consumer<JavaBlock> ifAction) {
        contents.tryBlock(resource, ifAction);
        return new JavaTryBlock(contents);
    }

    public final void lambda(String parameterType, String parameterName, Consumer<JavaLambda> body) {
        contents.lambda(parameterType, parameterName, body);
    }

    public final void lambda(String parameterType, String parameterName, String returnExpression) {
        contents.lambda(parameterType, parameterName, returnExpression);
    }
}
